File: src/linked-list/linked_list_intersection.go
package linkedlist

import "algorithm-puzzle-exercise/src/utils"

func LinkedListIntersection[T comparable](headA, headB *utils.ListNode[T]) *utils.ListNode[T] {
	if headA == nil || headB == nil {
		return nil
	}

	pA := headA
	pB := headB

	for pA != pB {
		if pA == nil {
			pA = headB
		} else {
			pA = pA.Next
		}

		if pB == nil {
			pB = headA
		} else {
			pB = pB.Next
		}
	}

	return pA
}

File: src/linked-list/linked-list-intersection.ts
import { ListNode, ListNodeOrNull } from '../utils/types'

export function linkedListIntersection<T>(
    headA: ListNode<T>,
    headB: ListNode<T>
): ListNodeOrNull<T> {
    let pA: ListNodeOrNull<T> = headA
    let pB: ListNodeOrNull<T> = headB

    while (pA !== pB) {
        // iterate A -> B combined list
        pA = pA === null ? headB : pA.next

        // iterate B -> A combined list
        pB = pB === null ? headA : pB.next
    }

    return pA
}
